### PR TITLE
Base Code for Course Edit and Delete

### PR DESCRIPTION
**Pull request by spencerj411**
As #29 _Course Deletion (8)_ is currently blocked, I created a new branch that contains base code for both #31 _Course Editing (5)_ (@Daniel-Jang) and #29 instead.
The idea was to render just one instance of the course table and have an 'Options' column where the user could choose to either edit or delete a course. 

Notes:
1. code contains placeholders (sample links and images)
2. the icons used were from http://webpublishing.anu.edu.au/web-style-guide/icons.php. However, we are expected to contact ANU styling first if we want to use their icons for the MVP.
3. ANU styling has no icons that could be used for 'Delete' (hence the weird choice of thumbs down). However, we could also contact them about this if we want.

Capture

### CODE REVIEW

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
I understand the reasoning behind using a 'thumbs down' picture, I don't think it's a good idea to have this in the final version. The layout of the pictures is also slightly different for me (See attached picture). I also think the newly created the newly created search function on the `/list/` page would be useful in this instance, and since they perform a very similar function, it may be better just to adapt that page and use it to find courses to edit/delete. Plus it means that we could have a consistent page for editing/removing existing courses, degrees, or subplans.

Edit Format

Also, just a note. You could continue to work on the actual deletion of courses themselves, as that will take some time. The only area you are blocked on is searching for course dependencies in subplans

**Comment by spencerj411** [comment]
I should’ve been clearer about a few things:
1. The thumbs down icon was chosen temporarily just to see how 2 icons would look like in the Options column. We could either contact ANU styling about creating a new delete icon or we could just have them as text. Thoughts?
2. I have been working on the non-blocked criteria of Course Deletion but my thought process was to pr the overlapping work between delete and edit so @Daniel-Jang could start working on edit concurrently.

Idk why the icons are positioned like that though, I used their css; I’ll have a look into it. I’ll also have a look at `list.html` and implement the search function once changes have been merged.

**Comment by JackSekoranja** [comment]
1. I think the preferred option here would be to find different pictures (outside ANU ones) for the edit and delete functions, or if we couldn't do that then just use text.
2. Fair enough

I also wasn't suggesting you implement the search function yourself, since it has a fair amount of overhead. I thought it might be easier to replace the edit form with the list form and have a single consistent one (Which would benefit code re-use). I was thinking we could separate the main menu in to 3-4 basic functions: 
- **Add New Template/Subplan/Course**
- **View Existing Template/Subplan/Course**
- **Edit Existing Template/Subplan/Course**
- **Output/Print Template**

Then the View/Edit forms could go to the link page, and from there they can re-direct to your specific edit or delete page

**Comment by spencerj411** [comment]
I like the sound of that.

**Comment by Daniel-Jang** [comment]
I'd also be happy to follow Jack's suggestion. What I also thought of is that we can remove the 'remove' button from the table list view and only have the edit button. The button can then lead to an edit page, which then instead has a button to delete the course or to commit new changes. 

That way, you don't have to worry about finding the right delete button.

Also, I reckon you should add text next to the edit icon ("Edit"), just so it is clear to the user.

**Comment by spencerj411** [comment]
That sounds good to me too. But that has brought up another point which is what about if the user wants to mass delete a set of courses/templates/subplans?

**Comment by spencerj411** [comment]
What about something similar to this? This way, it will take less clicks for the user to edit/delete. Also changed actions back to text, safer to have for the mean time.
image

**Comment by JackSekoranja** [comment]
Yeah I was thinking something similar. It may even be better to remove the `edit and delete` text from the side and instead just have two buttons at the bottom saying edit selected and delete selected. I think this is a better idea since course editing and deletion is not supposed to happen very often, so it may be in our interests to make it MORE difficult to do

**Comment by KavsHub** [comment]
Hey so I was looking at this and comparing to my issue of #32. 
In the long run we're going to have heaps of courses, subplans and templates which means the table would get very long. So future thinking/early brainstorming (both you and I should look into maybe multi-pages - i.e next page etc, or a search/filter box to help the user find the relevant stuff)

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: This is an exception, so this could be rephrased perhaps, or at least an exclamation mark?

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: Is this here for a reason?

File: cassdegrees/templates/managecourses.html
Comment (j-selby) [inline cassdegrees/templates/managecourses.html:26]: Spacing here?

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: Maybe verify these are integers? Else buggy code could potentially do unwanted stuff here (they are passed directly into the url below)

**Review by j-selby — CHANGES_REQUESTED** [review]
No major complaints.

**Comment by Daniel-Jang** [comment]
Uhh, James, you reviewed at an unfortunate time... I have completely rewritten everything...
Can everyone please check the most recent commit? It's hard to explain everything in words so yeah just run the program and see for yourselves :) (note: add, edit and delete functionality only works for courses at this stage)

With the most recent commit, I have also completed this ticket: https://github.com/cass-degrees/CASS-Degrees-Code/issues/31

File: cassdegrees/templates/managecourses.html
Comment (j-selby) [inline cassdegrees/templates/managecourses.html:13]: We can't do this action stuff through POST?

File: cassdegrees/templates/managecourses.html
Comment (j-selby) [inline cassdegrees/templates/managecourses.html:15]: IMO, this needs to not point to the origin url itself (there is a Referer header anyway, isn't there?), but as to the specific functionality that navigating from this page brings.

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: I think you need to add some documentation here about what the specific differences are for this. It wasn't immediately clear to me what this suggested.

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: Un-needed logging.

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py:299]: Might be a good idea to note if there is something in offered_sems that *isn't* handled.

**Review by JackSekoranja — APPROVED** [review]
Looks good and works nicely. I would like to replace the radio buttons with check boxes so that the user can select and delete multiple courses at once, but I'm happy for us to create an issue to do that later so we can get this functionality out

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: Not sure what happened here.

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: Same feedback from before - rewording or at the very least an exclamation mark for an error.

File: cassdegrees/templates/list.html
Comment (j-selby) [inline cassdegrees/templates/list.html]: Wait, this endif ends *after* the form? That might break cases where there is no content.  (i.e `<form>` is unmatched).

**Review by j-selby — CHANGES_REQUESTED** [review]

File: cassdegrees/templates/managecourses.html
Comment (Daniel-Jang) [reply cassdegrees/templates/managecourses.html:13]: Which line are you talking about? I don't understand what you are suggesting.

File: cassdegrees/templates/managecourses.html
Comment (Daniel-Jang) [reply cassdegrees/templates/managecourses.html:15]: What are your suggested changes? I'm not sure how to proceed from your comment.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: Added some comments explaining this.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: Forgot to remove them, now fixed.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py:299]: Not sure what you mean by something that isn't handled?

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: Whoops, me neither, fixed now.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: Altered message to be "Please select a course to edit!"

File: cassdegrees/templates/list.html
Comment (Daniel-Jang) [reply cassdegrees/templates/list.html]: Fixed, didn't see that one.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: Changed message.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: No, don't think so, deleted.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: That shouldn't matter too much, the id's can't be altered by users (they are hidden input fields) and a post request with an unusual id made by something outside our Django app will be rejected since they don't have the required csrf token.

File: cassdegrees/templates/managecourses.html
Comment (j-selby) [reply cassdegrees/templates/managecourses.html:13]: ` action="/manage_courses/?action={{ action }}"`
Why are we mixing GET and POST?

File: cassdegrees/templates/managecourses.html
Comment (j-selby) [reply cassdegrees/templates/managecourses.html:15]: Instead of `managecourses.html`, name this option what it actually *changes*. If we are hardcoding the names of webpages here, this will get messy quick.

File: cassdegrees/ui/views.py
Comment (j-selby) [reply cassdegrees/ui/views.py:299]: offered_sems can potentially contain 'semester 1', 'semester 2', and (while we aren't explicitly handling it, as per our user stories), 'summer'/etc. Would just be nice to know if something unexpected is in here.

Comment (Daniel-Jang) [reply cassdegrees/ui/views.py:299]: Since 'semester 1' and 'semester 2' are defined by the form (and nothing else), there is no way we will get something unexpected. Users can't interact with this either. I think your suggested checks are overkill.

File: cassdegrees/templates/managecourses.html
Comment (Daniel-Jang) [reply cassdegrees/templates/managecourses.html:13]: How are we mixing GET and POST? The form most certainly states that the method is post. Adding "?action=" is not an indicator that it is a GET request.

Also, while we can create a new hidden input with the action variable from managecourses.html, there is this one problem. From list.html, there is no way to determine which button the user will press on until they do and that 'submits' the form immediately. Therefore, there is no possible way to quickly change this hypothetical hidden input value to the one we want ("Add", "Edit" or "Delete") before the form is submitted. Hence, we will have to keep it this way.

File: cassdegrees/templates/managecourses.html
Comment (Daniel-Jang) [reply cassdegrees/templates/managecourses.html:15]: Yeah OK that's a good reason. Will do.

File: cassdegrees/templates/managecourses.html
Comment (j-selby) [reply cassdegrees/templates/managecourses.html:13]: Action in a URL is a [GET parameter](https://en.wikipedia.org/wiki/Query_string). Adding action in the URL there is by definition a query string, which is what is used for GET requests. This just seems a little weird to me.

Its not too big on an issue, but something to note.

**Review by j-selby — APPROVED** [review]

File: cassdegrees/templates/managecourses.html
Comment (Daniel-Jang) [reply cassdegrees/templates/managecourses.html:13]: Actually, according to HTTP specifications, it is valid to have a query string for POST requests:
https://stackoverflow.com/questions/14710061/is-it-valid-to-combine-a-form-post-with-a-query-string/14710450
https://www.w3.org/Protocols/rfc2616/rfc2616.html

But I did just find out that it is generally not a good idea, so you were right to point that line out. Unfortunately, we still have to leave it this way due to the button problem :(

Comment (j-selby) [reply cassdegrees/templates/managecourses.html:13]: Na, totally fair enough - just as long as this is noted, I personally have no issue (as you said, it isn't technically incorrect).

File: cassdegrees/ui/views.py
Comment (j-selby) [reply cassdegrees/ui/views.py:299]: Fair enough.

**Pull request merged**
